feat(replay): Add examples for custom sampling (#8035)
getsentry/sentry-docs: src/platforms/javascript/common/session-replay/understanding-sessions.mdx

```

In `session` mode, this will upload any pending recording data to Sentry. In `buffer` mode, this will upload any pending recording data to Sentry and then continue recording, the same as when an error is sampled with `replaysOnErrorSampleRate`.

## Examples of Custom Sampling

> ### Replays For Specific Users

Reviewer: ```suggestion
### Replays for Specific Users
```

```javascript
const replay = new Sentry.Replay();

/**
 * Initialize the Sentry SDK as normal.
 * `replaysSessionSampleRate` is set to "0" so that we have manual control over session-based replays
 * `replaysOnErrorSampleRate` is set to "1.0" to always record a replay when an error occurs
 */
Sentry.init({
  dsn: "...",
  replaysSessionSampleRate: 0,
  replaysOnErrorSampleRate: 1.0,
  integrations: [replay],
});

// ...
// Sometime after your users are authenticated, check if they are an employee
// If they are, we call `replay.start()` to start recording a replay for the employee
// ...

// Check if user is an employee, if so, always record a replay
if (loggedInUser.isEmployee) {
  // You can get a reference to the Sentry Replay integration like so:
  const replay = Sentry.getCurrentHub().getIntegration(Sentry.Replay);
  // or alternatively
  // const replay = Sentry.getCurrentHub().getClient().getIntegrationById('Replay');

  // Start the replay!

>   replay.start();

Reviewer 2: This will not actually work! As since an error sample rate is set, replay is enabled and we'll run into this branch in `start()`:

```js
 if (this._isEnabled && this.recordingMode === 'buffer') {
      throw new Error('Replay buffering is in progress, call `flush()` to save the replay');
    }
```

I think you'd either need to do this:

```js
await replay.stop();
replay.start();
```

_or_ do this:

```js
replay.flush(); // <-- convert the error session to a regular session and start recording it
```

}
```

### Replays on Specific URLs

> Here is an example of always recording a replay when a user lands on a specific URL. Maybe this specific page has been buggy and you would like to always record a replay when a user lands on it so that your developers have a better opportunity to debug. Using the [Navigation API](https://developer.mozilla.org/en-US/docs/Web/API/Navigation_API) we can listen to navigation events for a specific URL, start the replay recording, and then stop the replay when the user leaves the URL.

Reviewer: ```suggestion
You can also choose to record a replay every time a user lands on a specified URL. This can help developers debug problematic pages. To do this, use the [Navigation API](https://developer.mozilla.org/en-US/docs/Web/API/Navigation_API), which can listen to navigation events for a specific URL, start the replay recording, and then stop it when the user leaves the URL.
```

```javascript
const replay = new Sentry.Replay();

/**
 * Initialize the Sentry SDK as normal.
 * `replaysSessionSampleRate` is set to its default value of "0.1" so that only ~10% of users are sampled
 * `replaysOnErrorSampleRate` is set to its default value of "1.0" to always record a replay when an error occurs
 */
Sentry.init({
  dsn: "...",
  replaysSessionSampleRate: 0.1,
  replaysOnErrorSampleRate: 1.0,
  integrations: [replay],
});

// Listen to navigation events for when users land on our buggy URL, appropriately named: "/buggy-page/"
// and then start recording a replay.
navigation.addEventListener("navigate", (event) => {
  const url = new URL(event.destination.url);

  if (url.pathname.startsWith("/buggy-page/")) {
    // User navigates to the buggy page, let's start recording

>     replay.start();

Reviewer 2: Same as above, you cannot start without stopping first! Or just call `flush()`, I guess.

```

In `session` mode, this will upload any pending recording data to Sentry. In `buffer` mode, this will upload any pending recording data to Sentry and then continue recording, the same as when an error is sampled with `replaysOnErrorSampleRate`.

## Examples of Custom Sampling

There are ways to enable custom sampling if you're interested in tracking a particular action or group, for example, a specific segment of users or a problematic URL.

### Replays for Specific Users

You can set up replays to be recorded automatically for a particular user group. Here's an example where we want to record session-based replays for all users who are employees. The `user` object has an `isEmployee` field to signal employee status.

```javascript
const replay = new Sentry.Replay();

/**
 * Initialize the Sentry SDK as normal.
 *
 * `replaysSessionSampleRate` is set to its default value of "0.1" so that only ~10% of users are sampled
 * `replaysOnErrorSampleRate` is set to its default value of "1.0" to always record a replay when an error occurs
 */
Sentry.init({
  dsn: "...",
  replaysSessionSampleRate: 0.1,
  replaysOnErrorSampleRate: 1.0,
  integrations: [replay],
});

// ...

> // Sometime after your users are authenticated, check if they are an employee
> // If they are, we call `replay.flush()` to either flush the replay as normal if session-based replay, or if it is buffering due to error sampling, the flush will send the first segment of the replay.

Reviewer: ```suggestion
// After a user has been authenticated, check if they're an employee
// If they are, we'll call `replay.flush()` to either flush the replay as normal if it's a session-based replay, or if it's buffering due to error sampling, the flush will send the first segment of the replay.
```

// Since `replaysOnErrorSampleRate` is 1, we know that a replay has been buffered and `flush()` will flush the contents of that buffer. If both sample rates are < 1, then there is a chance that a replay has not been recording/buffering.
// In this case, you can check that `replay.getReplayId()` returns a value, if it does, it means replay is active and you can call `replay.flush()`, other call `replay.start()` to start recording a new replay.

> // ...

Author: How does this sound @mydea ?

Reviewer 2: it's not fully correct - the `replaysOnErrorSampleRate` only needs to be > 0. As long as this is > 0, you'll always have a buffer. So the instructions should be something like:

As long as `replaysOnErrorSampleRate` is greater than 0, we will always buffer the replay in the background. 
When you then call `flush()`, we will ensure to always send whatever has been buffered before.

It is also safe to call `flush()` in any time, really. If the replay is not enabled, it will simply do nothing. If the replay is currently in `session` mode, it will send what has been buffered, and if it is in `buffer` mode, it will also send what has been buffered _and_ convert the session to `session` mode, like when an error occurred.

  replaysSessionSampleRate: 0.1,
  replaysOnErrorSampleRate: 1.0,
  integrations: [replay],
});

// ...
// After a user has been authenticated, check if they're an employee
// If they are, we'll call `replay.flush()` to either flush the replay as normal if it's a session-based replay, or if it's buffering due to error sampling, the flush will send the first segment of the replay.

// Since `replaysOnErrorSampleRate` is > 0, we know that a replay has been buffered and `flush()` will flush the contents of that buffer. If only `replaysSessionSampleRate` is > 1, then there is a chance that a replay has not been recording/buffering.
// In this case, you can check that `replay.getReplayId()` returns a value, if it does, it means replay is active and you can call `replay.flush()`, other call `replay.start()` to start recording a new replay.
// ...

// Check if user is an employee, if so, always record a replay
if (loggedInUser.isEmployee) {
  // You can get a reference to the Sentry Replay integration like so:
  const replay = Sentry.getCurrentHub().getIntegration(Sentry.Replay);
  // or alternatively
  // const replay = Sentry.getCurrentHub().getClient().getIntegrationById('Replay');

>   // `replaysOnErrorSampleRate` is > 0, so a replay will always be buffered while it waits for an error to happen.
>   // `flush()` will create a new replay if it is in a buffering state, otherwise will flush the recording segment if a replay is currently in progress.

Reviewer 2: ```suggestion
```

I think this can be removed now, as this is described above?

```javascript
const replay = new Sentry.Replay();

/**
 * Initialize the Sentry SDK as normal.
 *
 * `replaysSessionSampleRate` and `replaysOnErrorSampleRate` are both set to
 * "0" so that we have manual control over session-based replays
 */
Sentry.init({
  dsn: "...",
  replaysSessionSampleRate: 0,
  replaysOnErrorSampleRate: 0,
  integrations: [replay],
});

// Listen to navigation events for when users land on our buggy URL, appropriately named: "/buggy-page/"
// and then start recording a replay.
navigation.addEventListener("navigate", (event) => {
  const url = new URL(event.destination.url);

  if (url.pathname.startsWith("/buggy-page/")) {
    // User navigates to the buggy page, let's start recording
    replay.start();
  } else if (window.location.pathname.startsWith("/buggy-page/")) {
    // Stop replay when the user leaves the buggy page

>     replay.stop();

Reviewer 2: I actually would not recommend that - if we started (not buffered!) there is no need to stop, IMHO it is better to just leave it recording in that case (you can disregard what was captured after the user moved away if you want).

Otherwise you need to implement awaiting on the `stop()`, ensuring stop has finished before you call `start` again etc.
